### PR TITLE
Move the GitHub Teams API

### PR DESCRIPTION
**Pull request by atorr**
Fixes: #1387.

### CODE REVIEW

**Comment by weeco** [comment]
I think this PR is missing (some?) of the new Team API endpoints e. g. fetching a team by it's slug `GET /orgs/:org/teams/:team_slug`, see: https://developer.github.com/changes/2020-01-21-moving-the-team-api-endpoints/

**Comment by gmlewis** [comment]
Thank you, @weeco - I've made a list in #1387 so we can hopefully keep track of the changes needed, and the ones that are implemented, as we review the PR(s).

File: github/teams.go [hunk 1]
```diff
@@ -239,11 +239,38 @@ func (s *TeamsService) EditTeam(ctx context.Context, id int64, team NewTeam, rem
 	return t, resp, nil
 }
 
-// DeleteTeam deletes a team.
+// EditTeamByName edits a team, given an organiation name, by name.
```
Comment (gmlewis) [inline github/teams.go]: For consistency, please rename this function to `EditTeamBySlug` and change comment to say "by slug."

File: github/teams.go [hunk 2]
```diff
@@ -252,11 +279,24 @@ func (s *TeamsService) DeleteTeam(ctx context.Context, team int64) (*Response, e
 	return s.client.Do(ctx, req, nil)
 }
 
-// ListChildTeams lists child teams for a team.
+// DeleteTeamByName deletes a team reference by Name.
+//
+// GitHub API docs: https://developer.github.com/v3/teams/#delete-team
+func (s *TeamsService) DeleteTeamByName(ctx context.Context, org, slug string) (*Response, error) {
```
Comment (gmlewis) [inline github/teams.go]: For consistency, please rename this function to `DeleteTeamBySlug` and change comment to day "reference by slug."

File: github/teams.go [hunk 3]
```diff
@@ -276,11 +316,63 @@ func (s *TeamsService) ListChildTeams(ctx context.Context, teamID int64, opt *Li
 	return teams, resp, nil
 }
 
-// ListTeamRepos lists the repositories that the specified team has access to.
+// ListChildTeamsByParentName lists child teams for a parent team given parent name.
```
Comment (gmlewis) [inline github/teams.go]: For consistency, please rename this function to `ListChildTeamsByParentSlug`, and change the comment to say "given parent slug".

File: github/teams.go [hunk 4]
```diff
@@ -276,11 +316,63 @@ func (s *TeamsService) ListChildTeams(ctx context.Context, teamID int64, opt *Li
...
+	var repos []*Repository
+	resp, err := s.client.Do(ctx, req, &repos)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return repos, resp, nil
+}
+
+// ListTeamReposByName lists the repositories given a team name that the specified team has access to.
```
Comment (gmlewis) [inline github/teams.go]: `ListTeamReposBySlug`... `given a team slug`.

File: github/teams.go [hunk 5]
```diff
@@ -304,13 +396,37 @@ func (s *TeamsService) ListTeamRepos(ctx context.Context, team int64, opt *ListO
 	return repos, resp, nil
 }
 
-// IsTeamRepo checks if a team manages the specified repository. If the
+// IsTeamRepoByID checks if a team, given it's ID, manages the specified repository. If the
```
Comment (gmlewis) [inline github/teams.go]: s/it's/its/

File: github/teams.go [hunk 6]
```diff
@@ -304,13 +396,37 @@ func (s *TeamsService) ListTeamRepos(ctx context.Context, team int64, opt *ListO
...
+	repository := new(Repository)
+	resp, err := s.client.Do(ctx, req, repository)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return repository, resp, nil
+}
+
+// IsTeamRepoByName checks if a team, given it's name, manages the specified repository. If the
```
Comment (gmlewis) [inline github/teams.go]: `IsTeamRepoBySlug`... `given its slug`.

File: github/teams.go [hunk 7]
```diff
@@ -418,12 +588,36 @@ func (s *TeamsService) ListTeamProjects(ctx context.Context, teamID int64) ([]*P
 	return projects, resp, nil
 }
 
-// ReviewTeamProjects checks whether a team has read, write, or admin
+// ReviewTeamProjectsByID checks whether a team, given it's ID, has read, write, or admin
```
Comment (gmlewis) [inline github/teams.go]: s/it's/its/

File: github/teams.go [hunk 8]
```diff
@@ -418,12 +588,36 @@ func (s *TeamsService) ListTeamProjects(ctx context.Context, teamID int64) ([]*P
...
+	projects := &Project{}
+	resp, err := s.client.Do(ctx, req, &projects)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return projects, resp, nil
+}
+
+// ReviewTeamProjectsByName checks whether a team, given it's ID, has read, write, or admin
```
Comment (gmlewis) [inline github/teams.go]: `ReviewTeamProjectsBySlug`... `given its slug`.

File: github/teams.go [hunk 9]
```diff
@@ -356,13 +472,43 @@ func (s *TeamsService) AddTeamRepo(ctx context.Context, team int64, owner string
 	return s.client.Do(ctx, req, nil)
 }
 
-// RemoveTeamRepo removes a repository from being managed by the specified
-// team. Note that this does not delete the repository, it just removes it
-// from the team.
+// AddTeamRepoByName adds a repository to be managed by the specified team given the team name.
```
Comment (gmlewis) [inline github/teams.go]: `AddTeamRepoBySlug`... `given the team slug`.

File: github/teams.go [hunk 10]
```diff
@@ -356,13 +472,43 @@ func (s *TeamsService) AddTeamRepo(ctx context.Context, team int64, owner string
...
+	u := fmt.Sprintf("organizations/%v/team/%v/repos/%v/%v", orgID, teamID, owner, repo)
+	req, err := s.client.NewRequest("DELETE", u, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.client.Do(ctx, req, nil)
+}
+
+// RemoveTeamRepoByName removes a repository from being managed by the specified
```
Comment (gmlewis) [inline github/teams.go]: `RemoveTeamRepoBySlug`...`given the team slug.`

File: github/teams.go [hunk 11]
```diff
@@ -394,11 +540,35 @@ func (s *TeamsService) ListUserTeams(ctx context.Context, opt *ListOptions) ([]*
...
+	var projects []*Project
+	resp, err := s.client.Do(ctx, req, &projects)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return projects, resp, nil
+}
+
+// ListTeamProjectsByName lists the organization projects for a team given the team name.
```
Comment (gmlewis) [inline github/teams.go]: `ListTeamProjectsBySlug`... `given the team slug.`

File: github/teams.go [hunk 12]
```diff
@@ -473,16 +667,14 @@ func (s *TeamsService) AddTeamProject(ctx context.Context, teamID, projectID int
 	return s.client.Do(ctx, req, nil)
 }
 
-// RemoveTeamProject removes an organization project from a team. An organization owner or
-// a team maintainer can remove any project from the team. To remove a project from a team
-// as an organization member, the authenticated user must have "read" access to both the team
-// and project, or "admin" access to the team or project.
-// Note: This endpoint removes the project from the team, but does not delete it.
+// AddTeamProjectByName adds an organization project to a team given the team name.
```
Comment (gmlewis) [inline github/teams.go]: `AddTeamProjectBySlug`...`given the team slug.`

File: github/teams.go [hunk 13]
```diff
@@ -494,77 +686,46 @@ func (s *TeamsService) RemoveTeamProject(ctx context.Context, teamID int64, proj
...
-	if err != nil {
-		return nil, resp, err
-	}
-	return groups, resp, err
+	return s.client.Do(ctx, req, nil)
 }
 
-// CreateOrUpdateIDPGroupConnections creates, updates, or removes a connection between a team
-// and an IDP group.
+// RemoveTeamProjectByName removes an organization project from a team given team name.
```
Comment (gmlewis) [inline github/teams.go]: `RemoveTeamProjectBySlug`...`given team slug.`

**Comment by atorr** [comment]
@weeco - I will go ahead and take a look again.  As for the particular endpoint you referenced the implementation for that should be in github/teams.go on line [119](https://github.com/atorr/go-github/blob/60a41cf33f4e1bed91fb1bf357e8905aed317433/github/teams.go#L119).   Is that what you are referring to?

File: github/teams.go [hunk 14]
```diff
@@ -494,77 +686,46 @@ func (s *TeamsService) RemoveTeamProject(ctx context.Context, teamID int64, proj
 	return s.client.Do(ctx, req, nil)
 }
 
-// IDPGroupList represents a list of external identity provider (IDP) groups.
```
Comment (gmlewis) [inline github/teams.go]: Why are lines 497 to the end of the file being deleted in this PR?
Oh, it looks like you are also trying to reorganize the files in this PR as well?
Sorry, but please handle file reorganization in a separate PR.

Attempting to address the changes listed in the announcement: https://developer.github.com/changes/2020-01-21-moving-the-team-api-endpoints/
while at the same time reorganizing files and moving code around makes this PR much much larger than it needs to be and is not reviewable as such.

So please keep this PR focused just on the checklist I made in #1387 and revert the rest of the changes and make a separate PR for those.  Thanks.

Comment (atorr) [reply github/teams.go]: Sure thing.  Must've missed the list.  Working on reverting the changes now.  Thanks!

**Comment by atorr** [comment]
New changes in.  Had missed the link to the blog post and went a bit overboard.  Will work to pay more attention to what needs to be changed.  Thank you all for your patience on this!

**Comment by weeco** [comment]
Is there any chance to get that merged/released soonish?

**Comment by gmlewis** [comment]
Hi @weeco - this repo is maintained by volunteers and it is difficult to tell when people will have time to review PRs, so I don't have an ETA for you.

I suggest that if you need something right away, you fork the repo and pull in the PR(s) that you need and work off of that.

Thank you for your patience.

**Comment by gmlewis** [comment]
Thank you, @wesleimp!
Merging.

**Comment by lynncyrin** [comment]
👋 @gmlewis I have a request 🙏 can you post a comment in this PR when there's a release that includes this change? I'd appreciate it tons ✨

**Comment by gmlewis** [comment]
@lynncyrin - https://github.com/google/go-github/releases/tag/v29.0.3 incorporates this change.

**Pull request merged**